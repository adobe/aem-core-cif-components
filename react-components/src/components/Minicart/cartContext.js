/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

import React, { createContext, useContext, useReducer } from 'react';
import { object, func } from 'prop-types';
import { useCookieValue } from '../../utils/hooks';

export const initialState = {
    isOpen: false,
    isRegistered: false,
    isEditing: false,
    isLoading: false,
    editItem: {},
    cartId: null,
    cart: null,
    errorMessage: null,
    couponError: null
};

export const reducerFactory = setCartCookie => {
    return (state, action) => {
        switch (action.type) {
            case 'close':
                return {
                    ...state,
                    isOpen: false
                };
            case 'open':
                return {
                    ...state,
                    isOpen: true
                };
            case 'beginLoading':
                return {
                    ...state,
                    isLoading: true
                };
            case 'endLoading':
                return {
                    ...state,
                    isLoading: false
                };
            case 'beginEditing':
                return {
                    ...state,
                    isEditing: true,
                    editItem: action.item
                };
            case 'endEditing':
                return {
                    ...state,
                    isEditing: false,
                    editItem: {}
                };
            case 'cartId':
                return {
                    ...state,
                    cartId: action.cartId
                };
            case 'reset':
                setCartCookie('', 0);
                return {
                    ...state,
                    cartId: null,
                    errorMessage: null,
                    couponError: null,
                    cart: null
                };
            case 'cart':
                return {
                    ...state,
                    cart: action.cart,
                    isLoading: false,
                    couponError: !action.cart.applied_coupon ? state.couponError : null
                };
            case 'register': {
                return {
                    ...state,
                    isRegistered: true
                };
            }
            case 'error':
                console.error(action.error);
                return {
                    ...state,
                    errorMessage: action.error
                };
            case 'discardError':
                return {
                    ...state,
                    errorMessage: null
                };
            case 'couponError':
                return {
                    ...state,
                    couponError: action.error
                };

            default:
                return state;
        }
    };
};

export const CartContext = createContext();

export const CartProvider = props => {
    const factory = props.reducerFactory || reducerFactory;
    const state = props.initialState || initialState;

    const [, setCartCookie] = useCookieValue('cif.cart');
    const contextValue = useReducer(factory(setCartCookie), state);

    return <CartContext.Provider value={contextValue}>{props.children}</CartContext.Provider>;
};

CartProvider.propTypes = {
    reducerFactory: func,
    initialState: object
};

export const useCartState = () => useContext(CartContext);
