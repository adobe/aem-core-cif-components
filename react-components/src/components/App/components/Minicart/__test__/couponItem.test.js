/*******************************************************************************
 *
 *    Copyright 2019 Adobe. All rights reserved.
 *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License. You may obtain a copy
 *    of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software distributed under
 *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 *    OF ANY KIND, either express or implied. See the License for the specific language
 *    governing permissions and limitations under the License.
 *
 ******************************************************************************/

import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { CartProvider } from '../cartContext';
import CouponItem from '../couponItem';

describe('<CouponItem />', () => {
    it('renders the component', () => {
        const initialState = {
            cart: {
                applied_coupon: {
                    code: 'my-sample-coupon'
                }
            }
        };

        const { asFragment } = render(
            <CartProvider initialState={initialState} reducerFactory={() => state => state}>
                <CouponItem />
            </CartProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('removes a coupon', () => {
        const mockFn = jest.fn();

        const initialState = {
            removeCoupon: mockFn,
            cart: {
                applied_coupon: {
                    code: 'my-sample-coupon'
                }
            }
        };

        const { getByText } = render(
            <CartProvider initialState={initialState} reducerFactory={() => state => state}>
                <CouponItem />
            </CartProvider>
        );

        fireEvent.mouseDown(getByText('Remove coupon'));
        expect(mockFn.mock.calls.length).toEqual(1);
    });
});
