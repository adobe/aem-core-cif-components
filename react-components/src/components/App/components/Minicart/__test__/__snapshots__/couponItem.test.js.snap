// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CouponItem /> renders the component 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="couponName"
    >
      Coupon 
      <strong>
        my-sample-coupon
      </strong>
       applied.
    </div>
    <div
      class="root"
    >
      <button
        class="kebab"
      >
        <span
          class="root"
        >
          <svg
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="1"
            />
            <circle
              cx="12"
              cy="5"
              r="1"
            />
            <circle
              cx="12"
              cy="19"
              r="1"
            />
          </svg>
        </span>
      </button>
      <ul
        class="dropdown"
      >
        <li
          class="menuItem"
        >
          <button>
            <span
              class="root"
            >
              <svg
                fill="none"
                height="14px"
                stroke="rgb(var(--venia-teal))"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="14px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polyline
                  points="3 6 5 6 21 6"
                />
                <path
                  d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
                />
              </svg>
            </span>
            <span
              class="text"
            >
              Remove coupon
            </span>
          </button>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
